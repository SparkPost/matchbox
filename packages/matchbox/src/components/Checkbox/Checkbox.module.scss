@import '../../styles/config';

.Checkbox {
  position: relative;
  border: none;
  padding: 0;
  margin: 0 0 spacing() 0;
  white-space: nowrap;

  &:last-child {
    margin: 0;
  }
}

.Input {
  @include visually-hidden;

  &:hover{
    & ~ .Control .Box {
      border: 1px solid color(gray, 5);
    }
  }

  &:focus ~ .Control .Box {
    box-shadow: 0 0 0 1px color(blue);
    border: 1px solid color(blue);
  }

  &:checked {
    & ~ .Control .Box {
      border: 1px solid color(blue);
    }

    & ~ .Control .Check {
      opacity: 1;
      transform: scale(1);
    }

    &:focus ~ .Control .Box,
    &:active ~ .Control .Box {
      box-shadow: 0 0 0 1px color(blue);
      border: 1px solid color(blue);
    }
  }

  &:disabled  ~ .Control .Box {
    background: color(gray, 9);
  }
  &:disabled  ~ .Control .Box,
  &:disabled  ~ .Control .Check {
    cursor: not-allowed;
  }
}

.Control {
  display: inline-block;
  vertical-align: top;
  &:hover {
    cursor: pointer;
  }
}

.Box, .Check {
  transition: 0.15s;
}

.Box {
  position: relative;
  top: 1px;
  left: 1px;
  width: 18px;
  height: 18px;
  border: 1px solid color(gray, 6);
  border-radius: border-radius();
}

.Check {
  position: absolute;
  top: 3px;
  left: 3px;
  stroke-width: 3px;
  stroke: color(blue);
  fill: color(blue);

  opacity: 0;
  transform: scale(0.5);
}

.Label {
  display: inline-block;
  vertical-align: top;
  padding-left: 9px;
  margin-top: 1px;
  line-height: 18px;
  white-space: nowrap;
}

.Group {
  display: block;
  margin-bottom: spacing();

  & > * {
    margin-bottom: 0;
  }

  &:last-child {
    margin: 0;
  }
}

.GroupLabel {
  padding-bottom: spacing(small);
}

.HelpText {
  font-size: font-size(300);
  line-height: line-height(300);
  color: color(gray, 4);
  padding-left: 27px;
}
