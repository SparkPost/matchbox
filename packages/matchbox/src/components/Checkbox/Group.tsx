import React from 'react';
import { Box } from '../Box';
import { Label } from '../Label';
import { Stack } from '../Stack';
import { OptionalLabel } from '../OptionalLabel';
import styled from 'styled-components';
import { margin, MarginProps } from 'styled-system';
import { pick } from '@styled-system/props';

const StyledGroup = styled('fieldset')`
  border: none;
  padding: 0;
  ${margin}
`;

type CheckboxGroupProps = {
  children: React.ReactNode;
  className?: string;
  'data-id'?: string;
  label?: React.ReactNode;
  required?: boolean;
  labelHidden?: boolean;
  optional?: boolean;
} & MarginProps;

function Group(props: CheckboxGroupProps): JSX.Element {
  const {
    'data-id': dataId,
    children,
    className,
    label,
    labelHidden,
    required,
    optional,
    ...rest
  } = props;
  const systemProps = pick(rest);

  return (
    <StyledGroup {...systemProps} className={className} data-id={dataId}>
      {label && (
        <Box width="100%">
          <Label as="legend" labelHidden={labelHidden}>
            <Box as="span" pr="200">
              {label}
            </Box>
            {required && (
              <Box as="span" pr="200" aria-hidden="true">
                *
              </Box>
            )}
            {optional && <OptionalLabel />}
          </Label>
        </Box>
      )}
      <Stack space="100">{children}</Stack>
    </StyledGroup>
  );
}

Group.displayName = 'Checkbox.Group';
export default Group;
