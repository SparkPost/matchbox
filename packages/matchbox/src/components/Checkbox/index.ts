export { default as Checkbox } from './Checkbox';
export * from './Checkbox';
export * from './Group';
