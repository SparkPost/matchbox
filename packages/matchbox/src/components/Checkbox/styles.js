import { tokens } from '@sparkpost/design-tokens';
import { Box } from '../Box';
import { visuallyHidden } from '../../styles/helpers';
import styled from 'styled-components';
import { margin } from 'styled-system';

export const Wrapper = styled('div')`
  ${margin}
`;

export const StyledBox = styled(Box)`
  border: 2px solid ${props => (props.error ? tokens.color_red_700 : tokens.color_gray_700)};
  transition: ${tokens.motionDuration_fast} ${tokens.motionEase_in_out};
  transition-property: border, background;
`;

export const StyledCheck = styled(Box)`
  opacity: 0;
  color: ${tokens.color_white};
  transition: opacity ${tokens.motionDuration_fast} ${tokens.motionEase_in_out};
`;

export const StyledLabel = styled('label')`
  display: flex;
  align-items: center;
  margin: 0;
  padding: 0;
  white-space: nowrap;
  user-select: none;

  &:hover {
    cursor: ${props => (props.disabled ? 'not-allowed' : 'pointer')};

    ${StyledBox} {
      ${props => (!props.disabled ? `border: 2px solid ${tokens.color_gray_800};` : '')}
    }
  }
`;

export const StyledInput = styled('input')`
  ${visuallyHidden}

  &:focus ~ span ${StyledBox} {
    box-shadow: 0 0 0 2px ${tokens.color_white}, 0 0 0 4px ${tokens.color_blue_700}
  }

  &:checked {
    & ~ span ${StyledBox} {
      border: 2px solid ${props => (props.error ? tokens.color_red_700 : tokens.color_blue_700)};
      background: ${props => (props.error ? tokens.color_red_700 : tokens.color_blue_700)};
    }

    & ~ span ${StyledCheck} {
      opacity: 1;
    }
  }

  &:disabled ~ span ${StyledBox} {
    background: ${tokens.color_gray_300};
    border: 2px solid ${tokens.color_gray_600};
  }

  &:disabled:checked ~ span ${StyledBox} {
    background: ${tokens.color_gray_600};
    border: 2px solid ${tokens.color_gray_600};
  }
`;
