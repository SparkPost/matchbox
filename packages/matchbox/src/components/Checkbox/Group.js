import React from 'react';
import PropTypes from 'prop-types';
import { Label } from '../Label';
import styles from './Checkbox.module.scss';

const Group = ({ children, label, required }) => (
  <div className={styles.Group}>
    {label && <Label className={styles.GroupLabel}>{label}{required && ' *'}</Label>}
    {children}
  </div>
);

Group.propTypes = {
  children: PropTypes.node.isRequired,
  label: PropTypes.node,
  required: PropTypes.bool
};

Group.displayName = 'Checkbox.Group';
export default Group;
