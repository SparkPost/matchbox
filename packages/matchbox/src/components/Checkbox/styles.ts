import { tokens } from '@sparkpost/design-tokens';
import { Box, BoxProps } from '../Box';
import { visuallyHidden } from '../../styles/helpers';
import styled from 'styled-components';
import { margin } from 'styled-system';
import { focusOutline } from '../../styles/helpers';

type CheckboxIndeterminateProp = {
  $indeterminate?: boolean;
};

type CheckboxDisabledProp = {
  $disabled?: boolean;
};

type CheckboxErrorProp = {
  $error?: boolean;
};

export const Wrapper = styled.div`
  ${margin}
`;

export const StyledCheck = styled(Box)<BoxProps>`
  opacity: 0;
  color: ${tokens.color_white};
  transition: opacity ${tokens.motionDuration_fast} ${tokens.motionEase_in_out};

  input:checked ~ span & {
    opacity: 1;
  }
`;

export const StyledIndeterminate = styled(Box)<BoxProps & CheckboxIndeterminateProp>`
  ${(props) => `
    opacity: ${props.$indeterminate ? '1' : '0'};
  `}
`;

export const StyledLabel = styled('label')<CheckboxDisabledProp>`
  display: flex;
  align-items: center;
  margin: 0;
  padding: 0;
  white-space: nowrap;
  user-select: none;
  cursor: ${(props) => (props.$disabled ? 'not-allowed' : 'pointer')};
`;

export const StyledBox = styled(Box)<
  BoxProps & CheckboxIndeterminateProp & CheckboxDisabledProp & CheckboxErrorProp
>`
  border: 2px solid ${(props) => (props.$error ? tokens.color_red_700 : tokens.color_gray_700)};
  transition: ${tokens.motionDuration_fast} ${tokens.motionEase_in_out};
  transition-property: border, background;

  ${focusOutline({ modifier: 'input:focus ~ span &' })}

  input:checked ~ span & {
    border: 2px solid ${(props) => (props.$error ? tokens.color_red_700 : tokens.color_blue_700)};
    background: ${(props) => (props.$error ? tokens.color_red_700 : tokens.color_blue_700)};
  }

  input:disabled ~ span & {
    background: ${tokens.color_gray_300};
    border: 2px solid ${tokens.color_gray_600};
  }

  input:disabled:checked ~ span & {
    background: ${tokens.color_gray_600};
    border: 2px solid ${tokens.color_gray_600};
  }

  ${StyledLabel}:hover input:not(:disabled) ~ span & {
    ${(props) =>
      !props.$disabled && !props.$indeterminate && !props.$error
        ? `border: 2px solid ${tokens.color_gray_800};`
        : ''}
  }

  ${StyledLabel}:hover input:checked:not(:disabled) ~ span & {
    ${(props) => (!props.$error ? `border: 2px solid ${tokens.color_blue_700};` : '')}
  }

  ${(props) =>
    props.$indeterminate
      ? `
        background: ${tokens.color_blue_700};
        border: 2px solid ${tokens.color_blue_700};
      `
      : null}
`;

// Issue with aria-checked types
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type InputProps = { 'aria-checked'?: any };
export const StyledInput = styled.input<InputProps>`
  ${visuallyHidden}
`;
