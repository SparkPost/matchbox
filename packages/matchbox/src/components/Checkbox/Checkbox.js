import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { Label } from '../Label';
import { Error } from '../Error';
import { Check } from '@sparkpost/matchbox-icons';
import Group from './Group';
import styles from './Checkbox.module.scss';

class Checkbox extends Component {
  static displayName = 'Checkbox';

  static Group = Group;

  static defaultProps = {
    required: false
  }

  static propTypes = {
    id: PropTypes.string,
    checked: PropTypes.bool,
    label: PropTypes.node,
    labelHidden: PropTypes.bool,
    disabled: PropTypes.bool,
    required: PropTypes.bool,
    value: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.string
    ]),
    error: PropTypes.string,
    onChange: PropTypes.func,
    onFocus: PropTypes.func,
    onBlur: PropTypes.func,
    helpText: PropTypes.node
  };

  render() {
    const {
      id,
      checked,
      label,
      labelHidden,
      disabled,
      required,
      error,
      value,
      onChange,
      onFocus,
      onBlur,
      helpText,
      ...rest
    } = this.props;

    const setClasses = classnames(
      styles.Checkbox,
      error && styles.error
    );

    const requiredIndicator = required
      ? ' *'
      : '';

    const labelMarkup = label && !labelHidden
      ? <Label id={id} className={styles.Label}>
        {label}{requiredIndicator}
      </Label>
      : null;

    const errorMarkup = error
      ? <Error error={error} />
      : null;

    const helpMarkup = helpText
      ? <div className={styles.HelpText}>{ helpText }</div>
      : null;

    return (
      <fieldset className={setClasses}>
        <input
          id={id}
          value={value}
          checked={checked}
          disabled={disabled}
          className={styles.Input}
          onChange={(event) => onChange(event)}
          onFocus={onFocus}
          onBlur={onBlur}
          type='checkbox'
          {...rest} />
        <label htmlFor={id} className={styles.Control}>
          <div className={styles.Box} />
          <Check className={styles.Check} size={14}/>
        </label>
        { labelMarkup }
        { errorMarkup }
        { helpMarkup }
      </fieldset>
    );
  }
}

export default Checkbox;
