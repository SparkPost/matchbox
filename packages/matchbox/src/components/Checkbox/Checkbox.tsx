import React from 'react';
import { Label } from '../Label';
import { Error } from '../Error';
import { Box } from '../Box';
import { Check } from '@sparkpost/matchbox-icons';
import Group from './Group';
import { HelpText } from '../HelpText';
import useInputDescribedBy from '../../hooks/useInputDescribedBy';
import { pick } from '@styled-system/props';
import { omit } from '../../helpers/props';
import { margin } from 'styled-system';
import {
  StyledLabel,
  StyledCheck,
  StyledBox,
  StyledInput,
  StyledIndeterminate,
  Wrapper,
} from './styles';

export type CheckboxProps = {
  id: string;
  checked?: boolean | 'indeterminate';
  defaultChecked?: boolean;
  disabled?: boolean;
  error?: React.ReactNode;
  helpText?: React.ReactNode;
  label?: React.ReactNode;
  labelHidden?: boolean;
  required?: boolean;
  value?: string;
} & Omit<React.ComponentPropsWithRef<'input'>, 'checked'>;

const Checkbox = React.forwardRef<HTMLInputElement, CheckboxProps>(function Checkbox(
  props,
  userRef,
) {
  const {
    id,
    checked,
    defaultChecked,
    disabled,
    error,
    helpText,
    label,
    labelHidden,
    value,
    ...rest
  } = props;

  const componentProps = omit(rest, margin.propNames);
  const systemProps = pick(rest);
  const isControlled = checked !== undefined;

  const { describedBy, errorId, helpTextId } = useInputDescribedBy({
    id,
    hasHelpText: !!helpText,
    hasError: !!error,
  });

  const isIndeterminate = typeof checked === 'string' && checked === 'indeterminate';
  const indeterminateAttributes = isIndeterminate
    ? { 'aria-checked': 'mixed', checked: false }
    : { 'aria-checked': checked ? 'true' : 'false', checked: Boolean(checked) };

  return (
    <Wrapper {...systemProps}>
      <StyledLabel htmlFor={id} $disabled={disabled}>
        <StyledInput
          aria-invalid={!!error}
          id={id}
          value={value}
          defaultChecked={defaultChecked}
          disabled={disabled}
          type="checkbox"
          ref={userRef}
          {...(isControlled ? indeterminateAttributes : {})}
          {...describedBy}
          {...componentProps}
        />
        <Box
          as="span"
          position="relative"
          flex="0"
          display="flex"
          alignItems="center"
          height="1.5rem" // Matches height of Label
        >
          <StyledBox
            as="span"
            display="inline-block"
            borderRadius="200"
            size="1rem"
            lineHeight="200"
            $error={!!error}
            $indeterminate={isIndeterminate}
          />
          <StyledCheck position="absolute" width="1rem" height="1rem" as={Check} />
          <StyledIndeterminate
            position="absolute"
            left="100"
            right="100"
            height="2px"
            bg="white"
            borderRadius="200"
            $indeterminate={isIndeterminate}
          />
        </Box>
        <Box flex="1" pl="200">
          {label && (
            <Label
              as="span" // Outer wrapper already includes a label
              id={id}
              labelHidden={labelHidden}
              fontWeight="400"
              mb="0" // TODO Remove once margin 0 is baked into Label
            >
              {label}
            </Label>
          )}
        </Box>
      </StyledLabel>
      {helpText && (
        <HelpText id={helpTextId} ml="500" mt="0">
          {helpText}
        </HelpText>
      )}
      {error && <Error id={errorId} error={error} ml="500" />}
    </Wrapper>
  );
}) as React.ForwardRefExoticComponent<CheckboxProps> & {
  Group: typeof Group;
};

Checkbox.displayName = 'Checkbox';
Checkbox.Group = Group;
export default Checkbox;
