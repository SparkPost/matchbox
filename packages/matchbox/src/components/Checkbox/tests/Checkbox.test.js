import React from 'react';
import Checkbox from '../Checkbox';
import { shallow } from 'enzyme';
import cases from 'jest-in-case';

describe('Checkbox', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Checkbox />);
  });

  const testCases = [
    { name: 'checked', props: { checked: true }},
    { name: 'labelHidden', props: { labelHidden: true }},
    { name: 'disabled', props: { disabled: true }},
    { name: 'required', props: { required: true, label: 'Required field' }},
    { name: 'with label', props: { label: 'Check me!' }},
    { name: 'label with labelHidden', props: { label: 'This should not be visible', labelHidden: true }},
    { name: 'with error', props: { error: 'Checkbox error' }},
    { name: 'with helptext', props: { helpText: <p>This is helptext for checkbox</p> }},
    { name: 'with id', props: { id: 'chkCheckbox' }}
  ];

  cases('renders checkbox states', (opts) => {
    const newProps = {};
    newProps[opts.prop] = opts.val;

    wrapper.setProps(opts.props);

    expect(wrapper).toMatchSnapshot();
  }, testCases);

});

