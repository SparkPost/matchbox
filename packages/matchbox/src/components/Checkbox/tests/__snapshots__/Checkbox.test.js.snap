// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders checkbox states checked 1`] = `
<fieldset
  className="Checkbox"
>
  <input
    checked={true}
    className="Input"
    type="checkbox"
  />
  <label
    className="Control"
  >
    <div
      className="Box"
    />
    <Check
      className="Check"
      size={14}
    />
  </label>
</fieldset>
`;

exports[`Checkbox renders checkbox states disabled 1`] = `
<fieldset
  className="Checkbox"
>
  <input
    className="Input"
    disabled={true}
    type="checkbox"
  />
  <label
    className="Control"
  >
    <div
      className="Box"
    />
    <Check
      className="Check"
      size={14}
    />
  </label>
</fieldset>
`;

exports[`Checkbox renders checkbox states label with labelHidden 1`] = `
<fieldset
  className="Checkbox"
>
  <input
    className="Input"
    type="checkbox"
  />
  <label
    className="Control"
  >
    <div
      className="Box"
    />
    <Check
      className="Check"
      size={14}
    />
  </label>
</fieldset>
`;

exports[`Checkbox renders checkbox states labelHidden 1`] = `
<fieldset
  className="Checkbox"
>
  <input
    className="Input"
    type="checkbox"
  />
  <label
    className="Control"
  >
    <div
      className="Box"
    />
    <Check
      className="Check"
      size={14}
    />
  </label>
</fieldset>
`;

exports[`Checkbox renders checkbox states required 1`] = `
<fieldset
  className="Checkbox"
>
  <input
    className="Input"
    type="checkbox"
  />
  <label
    className="Control"
  >
    <div
      className="Box"
    />
    <Check
      className="Check"
      size={14}
    />
  </label>
  <Label
    className="Label"
  >
    Required field
     *
  </Label>
</fieldset>
`;

exports[`Checkbox renders checkbox states with error 1`] = `
<fieldset
  className="Checkbox error"
>
  <input
    className="Input"
    type="checkbox"
  />
  <label
    className="Control"
  >
    <div
      className="Box"
    />
    <Check
      className="Check"
      size={14}
    />
  </label>
  <Error
    error="Checkbox error"
  />
</fieldset>
`;

exports[`Checkbox renders checkbox states with helptext 1`] = `
<fieldset
  className="Checkbox"
>
  <input
    className="Input"
    type="checkbox"
  />
  <label
    className="Control"
  >
    <div
      className="Box"
    />
    <Check
      className="Check"
      size={14}
    />
  </label>
  <div
    className="HelpText"
  >
    <p>
      This is helptext for checkbox
    </p>
  </div>
</fieldset>
`;

exports[`Checkbox renders checkbox states with id 1`] = `
<fieldset
  className="Checkbox"
>
  <input
    className="Input"
    id="chkCheckbox"
    type="checkbox"
  />
  <label
    className="Control"
    htmlFor="chkCheckbox"
  >
    <div
      className="Box"
    />
    <Check
      className="Check"
      size={14}
    />
  </label>
</fieldset>
`;

exports[`Checkbox renders checkbox states with label 1`] = `
<fieldset
  className="Checkbox"
>
  <input
    className="Input"
    type="checkbox"
  />
  <label
    className="Control"
  >
    <div
      className="Box"
    />
    <Check
      className="Check"
      size={14}
    />
  </label>
  <Label
    className="Label"
  >
    Check me!
  </Label>
</fieldset>
`;
